File: TDDMicroExercises/Java/TelemetrySystem/src/main/java/tddmicroexercises/telemetrysystem/TelemetryDiagnosticControls.java
package tddmicroexercises.telemetrysystem;

public class TelemetryDiagnosticControls
{
    private final String DiagnosticChannelConnectionString = "*111#";
    
    private final TelemetryClient _telemetryClient;
    private String _diagnosticInfo = "";

        public TelemetryDiagnosticControls()
        {
            _telemetryClient = new TelemetryClient();
        }
        
        public String getDiagnosticInfo(){
        	return _diagnosticInfo;
        }
        
        public void setDiagnosticInfo(String diagnosticInfo){
        	this._diagnosticInfo = diagnosticInfo;
        }
 
        public void CheckTransmission() throws Exception
        {
            _diagnosticInfo = "";

        _telemetryClient.Disconnect();

        int retryLeft = 3;
        while (_telemetryClient.getOnlineStatus() == false && retryLeft > 0)
        {
            _telemetryClient.Connect(DiagnosticChannelConnectionString);
            retryLeft -= 1;
        }
         
        if(_telemetryClient.getOnlineStatus() == false)
        {
            throw new Exception("Unable to connect.");
        }

        _telemetryClient.Send(TelemetryClient.DIAGNOSTIC_MESSAGE);
        _diagnosticInfo = _telemetryClient.Receive();
    }
}

File: Cpp/TelemetrySystem/source/TelemetryDiagnosticControls.cpp

#include "../include/TelemetryDiagnosticControls.h"

#include <stdexcept>

TelemetryDiagnosticControls::TelemetryDiagnosticControls() : 
     m_diagnosticChannelConnectionString( "*111#" )
    ,m_telemetryClient()
    ,m_diagnosticInfo()
{}
    
std::string TelemetryDiagnosticControls::getDiagnosticInfo()
{
    return m_diagnosticInfo;
}

void TelemetryDiagnosticControls::setDiagnosticInfo(std::string const& diagnosticInfo)
{
    m_diagnosticInfo = diagnosticInfo;
}

void TelemetryDiagnosticControls::checkTransmission() 
{
    m_diagnosticInfo = "";

    m_telemetryClient.disconnect();

    int retryLeft = 3;
    while (m_telemetryClient.getOnlineStatus() == false && retryLeft > 0)
    {
        m_telemetryClient.connect(m_diagnosticChannelConnectionString);
        retryLeft -= 1;
    }
     
    if(m_telemetryClient.getOnlineStatus() == false)
    {
        throw std::invalid_argument("Unable to connect");
    }

    m_telemetryClient.send(TelemetryClient::DIAGNOSTIC_MESSAGE);
    m_diagnosticInfo = m_telemetryClient.receive();
}
    